File: 3. Longest Substring Without Repeating Characters/solution.java
import java.util.HashMap; 
class Solution {
    public int lengthOfLongestSubstring(String s) {
        int result = 0, prev = 0;
        HashMap<Character, Integer> ht = new HashMap<>();
        for (int i = 0; i < s.length(); i++) {
            Integer c = ht.get(s.charAt(i));
            if (c != null) {
                if (i - prev > result) result = i - prev;
                prev = c + 1; 
                ht.clear();
                i = c;
            }
            else ht.put(s.charAt(i), i);
        }
        if (s.length() - prev > result) result = s.length() - prev;
        return result;
    }
}

File: 3. Longest Substring Without Repeating Characters/solution.cpp
class Solution {
public:
    int lengthOfLongestSubstring(string s) {
        int l = 0, r = 0;
        int ans = 0;
        int window[128] = {};
        while (r < s.size() && l <= r) {
            while (l <= r && window[s[r]] > 0) 
                window[s[l++]]--;
            window[s[r++]]++;
            ans = max(r - l, ans);
        }
        return ans;
    }
};